Source/WebCore/Modules/url-pattern/URLPattern.cpp

```diff
@@ -324,4 +332,107 @@ ExceptionOr<void> URLPattern::compileAllComponents(ScriptExecutionContext& conte
     return { };
 }
 
+// https://urlpattern.spec.whatwg.org/#url-pattern-match
+ExceptionOr<std::optional<URLPatternResult>> URLPattern::match(ScriptExecutionContext& context, std::variant<URL, URLPatternInput>&& input, std::optional<String>&& baseURLString) const
+{
+    URLPatternResult result;
+    String protocol, username, password, hostname, port, pathname, search, hash;
+
+    bool isURLPatternInit = false;
+    if (std::holds_alternative<URLPatternInput>(input))
+        isURLPatternInit = std::holds_alternative<URLPatternInit>(std::get<URLPatternInput>(input));
+
+    if (isURLPatternInit) {
+        URLPatternInit inputAsInit = std::get<URLPatternInit>(std::get<URLPatternInput>(input));
+
+        if (!inputAsInit.baseURL.isEmpty())
+            return Exception { ExceptionCode::TypeError, "Initial URLPatternInit input should not contain a base URL string."_s };
+
+        auto maybeResult = processInit(WTFMove(inputAsInit), BaseURLStringType::URL);
+        if (maybeResult.hasException())
+            return { std::nullopt };
+
+        URLPatternInit processedInit = maybeResult.releaseReturnValue();
+        protocol = processedInit.protocol;
+        username = processedInit.username;
+        password = processedInit.password;
+        hostname = processedInit.hostname;
+        port = processedInit.port;
+        pathname = processedInit.pathname;
+        search = processedInit.search;
+        hash = processedInit.hash;
+
+        result.inputs = Vector<URLPatternInput> { inputAsInit };
+
+    } else {
+        URL inputAsURL;
+
+        if (std::holds_alternative<URL>(input))
+            inputAsURL = std::get<URL>(input);
+        else {
+            if (baseURLString.has_value()) {
+                inputAsURL = URL(*baseURLString);
+
+                if (!inputAsURL.isValid())
+                    return { std::nullopt };
+            }
+        }
+
+        ASSERT(!inputAsURL.isEmpty() && inputAsURL.isValid());
+
+        protocol = inputAsURL.protocol().toString();
+        username = inputAsURL.user();
+        password = inputAsURL.password();
+        hostname = !inputAsURL.host().isEmpty() ? inputAsURL.host().toString() : emptyString();
+        port = inputAsURL.port().has_value() ? String::number(*inputAsURL.port()) : emptyString();
```

Comment:
`has_value()` is usually unnecessary when checking if `std::optional` contains a value